### PR TITLE
Llama3.1 with torchtune

### PR DESCRIPTION
**Pull request by Gasoonjia**
This PR aims to add torchtune llama3.1 support while keep the original torchchat llama3.1 for reference.
To play with it, plz fetch it into a linxu machine, and uses [this script](https://gist.github.com/Gasoonjia/1a695d2be55aef723f7b0a84cf1ed62c)  to replace the current install/install_requirement.sh for environment installation.

The command for original model should be the same:
```
python3 torchchat.py generate llama3.1 --prompt "write me a story about a boy and his bear"
```
If you want to play with torchtune 3.1 model, consider using:
```
python3 torchchat.py generate llama3.1-tune --prompt "write me a story about a boy and his bear"
```

### CODE REVIEW

File: torchchat/cli/builder.py [hunk 1]
```diff
@@ -35,6 +35,14 @@
 from torchchat.utils.measure_time import measure_time
 from torchchat.utils.quantize import quantize_model
 
+# bypass the import issue before torchao is ready on macos
+try:
+    from torchtune.training import set_default_dtype
```
Comment (Jack-Khuu) [inline torchchat/cli/builder.py:40]: Unused import?

File: torchchat/model.py [hunk 2]
```diff
@@ -11,6 +11,7 @@
 from enum import Enum
 from pathlib import Path
 from typing import Callable, Dict, Optional, Union
+from abc import ABC, abstractmethod
```
Comment (Jack-Khuu) [inline torchchat/model.py:14]: ABC is unused?

File: torchchat/model.py [hunk 3]
```diff
@@ -290,50 +301,43 @@ def build_model(self) -> nn.Module:
         recipe = ModelRecipe.get_recipe(self.config.model_type)
         modules = {}
         for name, module_class in recipe.modules.items():
-            modules[name] = module_class(**self.config.transformer_args[name])
-        
+            if isinstance(self.config.transformer_args[name], dict):
+                modules[name] = module_class(**self.config.transformer_args[name])
+            else:
+                modules[name] = module_class(self.config.transformer_args[name])
```
Comment (Jack-Khuu) [inline torchchat/model.py:307]:
```suggestion
            if isinstance(config_args := self.config.transformer_args[name], dict):
                modules[name] = module_class(**config_args)
            else:
                modules[name] = module_class(config_args)
```

File: torchchat/model.py [hunk 4]
```diff
@@ -345,6 +349,44 @@ def from_gguf(cls, gguf_path: str, **kwargs):
...
+
+    def setup_caches(self, max_batch_size, dtype):
+        self.model.setup_caches(max_batch_size, dtype=dtype)
+
+    def reset_caches(self):
+        self.model.reset_caches()
+
+
+class FlamingoModel(Model):
+    def forward(self, tokens: Tensor, encoder_input: Optional[Dict[str, Tensor]] = None, encoder_mask: Optional[Tensor] = None) -> Tensor:
```
Comment (Jack-Khuu) [inline torchchat/model.py:372]: lint long line

**Review by Jack-Khuu — APPROVED** [review]
PR itself looks find to me with some minor nits.

File: torchchat/model.py (hunk 2 above)
Comment (Gasoonjia) [reply torchchat/model.py:14]: should be one of the Model's parents. Fixed it.

**Pull request merged**